### PR TITLE
apply php 7.2 to travis

### PR DESCRIPTION
**Pull request by samsonasik**

### CODE REVIEW

**Comment by samsonasik** [comment]
Got following errrors under php 7.2: 
```
There were 6 errors:

1) ZendTest\Session\Config\SessionConfigTest::testSavePathCanBeNonDirectoryWhenSaveHandlerNotFiles

Zend\Session\Exception\InvalidArgumentException: Invalid save handler specified: ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:148

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:68

2) ZendTest\Session\Config\SessionConfigTest::testSaveHandlerIsMutable

ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:97

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:147

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:912

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:105

3) ZendTest\Session\Config\SessionConfigTest::testSaveHandlerAltersIniSetting

ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:97

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:147

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:912

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:111

4) ZendTest\Session\Config\SessionConfigTest::testSetOptionSetsIniSetting with data set #2 ('save_handler', 'getOption', 'user')

ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:97

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:972

5) ZendTest\Session\Config\SessionConfigTest::testSetOptionsTranslatesUnderscoreSeparatedKeys with data set #2 ('save_handler', 'getOption', 'user')

ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:97

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:147

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:116

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:1005

6) ZendTest\Session\SaveHandler\MongoDBTest::testReadDestroysExpiredSession

ini_set(): Headers already sent. You cannot change the session module's ini settings at this time

/home/travis/build/zendframework/zend-session/test/SaveHandler/MongoDBTest.php:97
```

File: .travis.yml [hunk 1]
```diff
@@ -51,17 +51,17 @@ matrix:
...
+    - php: 7.2
       env:
         - DEPS=locked
-    - php: nightly
+    - php: 7.2
       env:
         - DEPS=latest
   allow_failures:
-    - php: nightly
+    - php: 7.2
```
Comment (michalbundyra) [inline .travis.yml]: Not sure if I am right, but somewhere, someone secretly decided to not allow failures on PHP 7.2 any more ;-)
@froschdesign Is it right? :smile:

Comment (froschdesign) [reply .travis.yml]: @samsonasik and @webimpress 
November 30, 2017 is around the corner therefore it make no sense! :smiley:

Comment (samsonasik) [reply .travis.yml]: @webimpress @froschdesign I've removed allow_failures->php 7.2, do you think we need to keep the "nightly"?

Comment (michalbundyra) [reply .travis.yml]: @samsonasik I wouldn't do that as long as we don't have PHP 7.3-alpha yet.

**Comment by samsonasik** [comment]
So, any suggestion to get the error gone in php 7.2 ?

**Comment by michalbundyra** [comment]
@samsonasik I'd suggest to try [`session_set_save_handler()`](http://php.net/manual/en/function.session-set-save-handler.php) method, because error message says:

```
Cannot set 'user' save handler by ini_set() or session_module_name()
```

**Comment by michalbundyra** [comment]
@samsonasik Hm... oh, no... sorry. No idea for now. This is not gonna work.

**Comment by remicollet** [comment]
Same issue from Fedora QA
https://apps.fedoraproject.org/koschei/package/php-zendframework-zend-session?collection=f28

**Comment by weierophinney** [comment]
This is a case where the tests are incorrect; the value "user" was never correct, but previous versions of PHP did not raise an error.

I'll update to provide a mock `SessionHandlerInterface` implementation.

File: src/Config/SessionConfig.php [hunk 2]
```diff
@@ -126,11 +136,31 @@ public function getStorageOption($storageOption)
...
+     * Prevents calls to `setSaveHandler()` from hitting `setOption()` instead,
+     * and thus bypassing the logic of `setPhpSaveHandler()`.
+     *
+     * @param  string $phpSaveHandler
+     * @return SessionConfig
+     * @throws Exception\InvalidArgumentException
+     */
+    public function setSaveHandler($phpSaveHandler)
+    {
+        // error_log(sprintf('In %s with %s', __METHOD__, var_export($phpSaveHandler, true)));
```
Comment (michalbundyra) [inline src/Config/SessionConfig.php:160]: Do we need this comment above?

File: src/SessionManager.php [hunk 3]
```diff
@@ -126,13 +127,23 @@ public function start($preserveStorage = false)
         $oldSessionData = [];
         if (isset($_SESSION)) {
             $oldSessionData = $_SESSION;
+
+            // convert session data to plain array that’ll be acceptable as
+            // ArrayUtils::merge parameter
+            if ($oldSessionData instanceof Storage\StorageInterface) {
+                $oldSessionData = $oldSessionData->toArray();
+            } elseif ($oldSessionData instanceof \Traversable) {
```
Comment (michalbundyra) [inline src/SessionManager.php]: `Traversable` is imported

File: src/SessionManager.php [hunk 4]
```diff
@@ -126,13 +127,23 @@ public function start($preserveStorage = false)
...
+                $oldSessionData = iterator_to_array($oldSessionData);
```
Comment (michalbundyra) [inline src/SessionManager.php]: I'm not sure if it is sufficient - because it is not recursive.

File: src/SessionManager.php [hunk 5]
```diff
@@ -126,13 +127,23 @@ public function start($preserveStorage = false)
...
+            }
         }
 
         session_start();
 
-        if ($oldSessionData instanceof \Traversable
-            || (! empty($oldSessionData) && is_array($oldSessionData))
-        ) {
+        if ($oldSessionData instanceof Traversable) {
```
Comment (michalbundyra) [inline src/SessionManager.php]: As noted by @Xerkus in separate PR this is no longer needed

File: src/SessionManager.php [hunk 6]
```diff
@@ -126,13 +127,23 @@ public function start($preserveStorage = false)
...
+            $oldSessionData = $oldSessionData->toArray();
+        }
+
+        if (! empty($oldSessionData) && is_array($oldSessionData)) {
```
Comment (michalbundyra) [inline src/SessionManager.php]: `! empty` is useless

File: src/Config/SessionConfig.php (hunk 2 above)
Comment (weierophinney) [reply src/Config/SessionConfig.php:160]: I'll remove; thanks!

**Comment by weierophinney** [comment]
Single test failure is on 5.6-locked, where the MongoDb SaveHandler implementation's garbage collection test reliably fails; passes on all other environments, however, so I'm going to take it as a pass.

**Comment by weierophinney** [comment]
Thanks, @samsonasik, for getting this started!

**Pull request closed without merging**